### ISSUE TITLE
cmd/compile: avoid call to runtime.makeslice for zero-length types?

### ISSUE THREAD

**Issue by bradfitz** [golang/go#65685]
### Go version

Go 1.22

### Output of `go env` in your module/workspace:

```shell
.
```

### What did you do?

`return make([]struct{}, 123)` compiles to a call to `runtime.makeslice`:

```
        LEAQ    type:struct {}(SB), AX
        MOVL    $123, BX
        MOVQ    BX, CX
        PCDATA  $1, $0
        NOP
        CALL    runtime.makeslice(SB)
        MOVL    $123, BX
        MOVQ    BX, CX
        ADDQ    $24, SP
        POPQ    BP
        RET
```

We see this show up in profiles (small, but non-existent) for calls to https://pkg.go.dev/tailscale.com/types/views#Slice.LenIter (which predates Go 1.22's range-over-int).

I realize that the compiler could omit those `makeslice` calls when the width of the type being made is zero (as there's no actual allocation).

(The compiler already does that when the slice length is zero.)

/cc @golang/compiler 

### What did you see happen?

call to makeslice

### What did you expect to see?

No call to runtime.makeslice.

/cc @twitchyliquid64 @maisem

**Comment by Jorropo** [comment]
I added this exact optimization for zero sized slices, I didn't added it for zero sized types since I couldn't figure out how to recover the `*types.Type` for a type compiled pointer in the SSA passes and this wasn't what I was trying to optimize so I never investigated it.

**Comment by Jorropo** [comment]
We could do this optimization in `genssa` but it's tricky because we can't rely on `prove` to properly prove that the indexes are correctly in bounds. Because this needs to panic:
```go
var a = -1
_ = make([]struct{}, a)
```

During `genssa` we can check if the arguments are literals or constants but anything else is very hard fast.
```go
for i := range make([]struct{}, 10) { // this would be optimized
 for j := range make([]struct{}, i) { /* ... */ } // here optimizing it in genssa would fail while SSA can use prove to handle this
}
```
Going back to your example:
```go
func (v Slice[T]) LenIter() []struct{} { return make([]struct{}, len(v.ж)) }
```
This is wouldn't be optimized in `genssa` as before prove we don't know what `len(v.ж)` could be.
I don't even know if `prove` would even properly handle this because it fails on some edge cases with `len` and `makeslice`.
---
Imo this should be handled in a third party linter which finds all the ranges of `make([]zeroSizedType, n)` calls and update the code to use range over int, `git grep -F "make([]struct{}, "` should get you 95% of the way there already.

**Comment by mauri870** [comment]
Regarding the SSA rule, I have come up with this:

```go
(StaticLECall {callAux} (Addr {t} _) _ _ mem)
	&& isSameCall(callAux, "runtime.makeslice")
	&& t.(*obj.LSym).String() == "type:struct {}"
	=> (MakeResult (Addr <v.Type.FieldType(0)> {ir.Syms.Zerobase} (SB)) mem)
```

But I have to admit that I lack proper understanding of this, there are probably better ways to achieve the same, if it is even possible this early in the optimization passes.

**Comment by mauri870** [comment]
I agree, my approach is very naive. There are probably lots of problems, I'm happy if it serves as a starting point somehow. Thanks for looking into this.

**Comment by Jorropo** [comment]
I will make a patch which implements this in `genssa` which would handle literals, constants and `len` / `cap` calls (I just realised they can't return an invalid length unless you use `unsafe` which is then your fault) as arguments to `make` if the compiler team think this a suitable (cc @randall77 ).
But imo consumers can rewrite their code to use range over int instead.

**Comment by randall77** [comment]
Yes, that's fine.
We could do it for all `make` calls, generating the negative test on the `make` argument and calling `runtime.panicmakeslicelen` if so. Not sure if it is worth it.

**Comment by Jorropo** [comment]
I thought we should inline hot code paths not cold ones.
